server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java

```diff
@@ -37,24 +37,42 @@ public BranchDao(System2 system2) {
   }
 
   public void insert(DbSession dbSession, BranchDto dto) {
+    setKeyType(dto);
     mapper(dbSession).insert(dto, system2.now());
   }
 
   public void upsert(DbSession dbSession, BranchDto dto) {
     BranchMapper mapper = mapper(dbSession);
     long now = system2.now();
+    setKeyType(dto);
     if (mapper.update(dto, now) == 0) {
       mapper.insert(dto, now);
     }
   }
 
+  private static void setKeyType(BranchDto dto) {
+    if (dto.getBranchType() == BranchType.PULL_REQUEST) {
+      dto.setKeyType(KeyType.PULL_REQUEST);
+    } else {
+      dto.setKeyType(KeyType.BRANCH);
+    }
+  }
+
   public int updateMainBranchName(DbSession dbSession, String projectUuid, String newBranchKey) {
     long now = system2.now();
     return mapper(dbSession).updateMainBranchName(projectUuid, newBranchKey, now);
   }
 
-  public Optional<BranchDto> selectByKey(DbSession dbSession, String projectUuid, String key) {
-    return Optional.ofNullable(mapper(dbSession).selectByKey(projectUuid, key));
+  public Optional<BranchDto> selectBranchByKey(DbSession dbSession, String projectUuid, String key) {
+    return selectByKey(dbSession, projectUuid, key, KeyType.BRANCH.name());
+  }
+
+  public Optional<BranchDto> selectPullRequestByKey(DbSession dbSession, String projectUuid, String key) {
```

Comment:
should we say pull request id or key? I think we should be consistent over the whole code and API